src/server/gc_worker/gc_worker.rs

```diff
@@ -247,40 +252,128 @@ struct KeysInRegions<R: Iterator<Item = Region>> {
+) -> Result<Box<dyn Iterator<Item = (Box<dyn Iterator<Item = Key>>, Region)>>> {
+    assert!(!keys.is_empty());
+
+    match region_or_provider {
+        Either::Left(region) => Ok(Box::new(KeysInRegions {
+            keys: keys.into_iter().peekable(),
+            regions: vec![region].into_iter().peekable(),
+        })),
+        Either::Right(region_provider) => {
+            if keys.len() >= 2 {
+                let start = keys.first().unwrap().as_encoded();
+                let end = keys.last().unwrap().as_encoded();
+                let regions = box_try!(region_provider.get_regions_in_range(start, end))
+                    .into_iter()
+                    .filter(move |r| find_peer(r, store_id).is_some())
+                    .peekable();
+
+                let keys = keys.into_iter().peekable();
+                Ok(Box::new(KeysInRegions { keys, regions }))
+            } else {
+                // We only have one key.
+                let key = keys.first().unwrap().as_encoded();
+                let region = box_try!(region_provider.find_region_by_key(key));
+
+                if region.is_none() {
+                    // Return a empty iterator when region has not been found.
+                    return Ok(Box::new(KeysInRegions {
+                        keys: Vec::new().into_iter().peekable(),
+                        regions: Vec::new().into_iter().peekable(),
+                    }));
+                }
 
-            let keys = keys.into_iter().peekable();
-            return Ok(Box::new(KeysInRegions { keys, regions }));
+                Ok(Box::new(KeysInRegions {
+                    keys: keys.into_iter().peekable(),
+                    regions: vec![region.unwrap()].into_iter().peekable(),
+                }))
+            }
         }
     }
-    Ok(Box::new(keys.into_iter()))
+}
+
+fn seek_region(
+    store_id: u64,
+    key: &[u8],
+    region_provider: Arc<dyn RegionInfoProvider>,
+) -> Result<Region> {
+    let (tx, rx) = mpsc::channel();
+    box_try!(region_provider.seek_region(
+        key,
+        Box::new(move |iter| {
+            for info in iter {
+                if find_peer(&info.region, store_id).is_some() {
+                    let _ = tx.send(Some(info.region.clone()));
+                    return;
+                }
+            }
+            let _ = tx.send(None);
+        }),
+    ));
+
+    match box_try!(rx.recv()) {
+        Some(region) => Ok(region),
+        None => unreachable!(),
+    }
+}
+
+fn init_snap_ctx(store_id: u64, region: &Region) -> Context {
+    let mut ctx = Context::default();
+    ctx.region_id = region.id;
+    ctx.region_epoch = region.region_epoch.clone();
+    ctx.stale_read = true;
+
+    let peers: Vec<_> = region
+        .peers
+        .iter()
+        .filter(|peer| peer.store_id == store_id)
+        .collect();
+
+    if peers.len() == 1 {
+        ctx.peer = SingularPtrField::some(peers[0].clone());
```

Comment:
You mean `ctx.set_peer(peers[0].clone())`?

Reply:
yes